documentation/src/docs/watermarking.md

```diff
@@ -0,0 +1,192 @@
+2. Generating consolidated watermarks.
+
+Writers may contact any arbitrary Controller instance and report their `timestamps` and `Positions`. This Controller conditionally persists the reported value in the shared metadata store. For generating watermarks, all available Controller instances divide ownership of Streams amongst themselves such that each Controller instance is exclusively responsible for computing watermarks for only a subset of Streams. So for any Stream there is exactly for one Controller instance which is periodically generating marks by consolidating all the individual Writer marks which could be processed by any arbitrary Controller instance.
+
+### Writer Marks
+There are two types of Writers (Event Writers and Transaction Writers) each with their own characteristics with respect to time to position tracking. So Event Writers and Transaction Writers will have their corresponding marks generated differently. However, their marks will be reported at the same place in the shared metadata store described above.
+
+- **Event Writer Marks**: Event Writers are expected to supply their positions and corresponding times explicitly to any arbitrary Controller instance. This Controller instance will durably record the received mark in the store after checking the mark for sanity.
+
+- **Transaction Writer Marks**: Transaction writers, on the other hand, are only aware of Transaction ids and the Transaction commits happen asynchronously and are orchestrated by Controller. Since Controller knows when a Transaction is committed, it is best suited to compute the "position" where it was committed.
+
+### Watermark Computation
+A watermark is composed of a time **T** and `streamCut` **S**. A watermark effectively indicates that "until `streamCut` **S**, all data corresponding to time **T** has been written across all Writers". To compose such a watermark from the given positions of Writers, we need to take the **lowest** time from marks across Writers and compute a greatest upper bound on all positions from across Writers marks. To compute greatest upper bound on, we will loop over each Writer's mark and loop over each Segment in Writer's position and include it in the set conditionally - a Segment is included in upper bound only if none of its successors already exist in the upper bound. An included Segment will replace all its predecessor Segments in the result set. A Segment included in the upper bound will include the highest offset across all Writer marks that contain that Segment.
+
+It is important to understand that the Writer's position is not same as a `streamCut` because a Writer could be writing concurrently to a predecessor and successor Segment but for different Routing Keys. This means any composition of Writer positions may not yield a well formed and complete `streamCut`. So the above computed greatest upper bound may have missing key ranges whereby we will need to augment above algorithm to compute a complete `streamCut`.
```

Comment:
"whereby we will need to augment above algorithm to compute a complete `streamCut`" - That makes it sound like its saying the feature is not implemented correctly.